packages/store/src/internal/state-factory.ts

```diff
@@ -49,83 +50,98 @@ export class StateFactory {
     private _parentFactory: StateFactory,
     private _actions: InternalActions,
     private _actionResults: InternalDispatchedActionResults,
-    private _stateContextFactory: StateContextFactory
+    private _stateContextFactory: StateContextFactory,
+    private _internalStateOperations: InternalStateOperations
   ) {}
 
-  /**
-   * Add a new state to the global defs.
-   */
-  add(oneOrManyStateClasses: StateClass | StateClass[]): MappedStore[] {
-    let stateClasses: StateClass[];
-    if (!Array.isArray(oneOrManyStateClasses)) {
-      stateClasses = [oneOrManyStateClasses];
+  private _states: MappedStore[] = [];
+
+  public get states(): MappedStore[] {
+    return this._parentFactory ? this._parentFactory.states : this._states;
+  }
+
+  public get rooStateNames(): Map<StateName, StateClassName> {
+    return this._parentFactory ? this._parentFactory.rooStateNames : this._statesNames;
+  }
+
+  public set rooStateNames(names: Map<StateName, StateClassName>) {
+    const stateNames: Map<StateName, StateClassName> = this.rooStateNames;
+    for (const [key, value] of names) {
+      stateNames.set(key, value);
+    }
+  }
+
+  private get stateTreeRef(): ObjectKeyMap<any> {
+    return this._internalStateOperations.getRootStateOperations().getState();
+  }
+
+  private static immutableDefaults(defaults: any): any {
+    let value = {};
+
+    if (Array.isArray(defaults)) {
+      value = [...defaults];
+    } else if (isObject(defaults)) {
+      value = { ...defaults };
+    } else if (defaults === undefined) {
```

Comment:
do you need this condition?

Reply:
Why not?

Reply:
i mean you declare a `value` variable before all these conditions and assign to `{}` and also in this condition you assign to `{}`, maybe better:
```TS
let value = {};

if (Array.isArray(defaults)) {
  value = [...defaults];
} else if (isObject(defaults)) {
  value = { ...defaults };
} else if (defaults !== undefined) {
  value = defaults;
}
```

Reply:
Just my 2c, I would say that having multiple returns would actually be better here. 
Guard clauses are one of the definite exceptions to the guidelines from Code Complete.
Could be expressed as:
```TS
if (Array.isArray(defaults)) return [...defaults];
if (isObject(defaults)) return { ...defaults };
if (typeof defaults === 'undefined') return {};
return defaults;
```